parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,337 @@
+            self.metrics = SharedTable(self.metrics)
+            self.model.share_memory()
+        shared['metrics'] = self.metrics
+
+        # TODO: Consider moving shared candidates to shared memory?
+        shared['fixed_candidates'] = self.fixed_candidates
+        shared['fixed_candidates_vec'] = self.fixed_candidates_vec
+        return shared
+
+    def update_params(self):
+        """Do optim step and clip gradients if needed."""
+        if self.clip > 0:
+            torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.clip)
+        self.optimizer.step()
+
+    def reset_metrics(self):
+        """Reset metrics."""
+        super().reset_metrics()
+        self.metrics['loss'] = 0.0
+        self.metrics['examples'] = 0
+        self.metrics['rank'] = 0
+
+    def report(self):
+        """Report loss and mean_rank from model's perspective."""
+        m = {}
+        examples = self.metrics['examples']
+        if examples > 0:
+            m['examples'] = examples
+            if self.metrics['loss'] > 0:
+                m['loss'] = self.metrics['loss']
+                m['mean_loss'] = self.metrics['loss'] / examples
+            if self.metrics['rank'] > 0:
+                m['mean_rank'] = self.metrics['rank'] / examples
+        for k, v in m.items():
+            # clean up: rounds to sigfigs and converts tensors to floats
+            m[k] = round_sigfigs(v, 4)
+        return m
+
+    def _get_model_file(self, opt):
+        model_file = None
+
+        # first check load path in case we need to override paths
+        if opt.get('init_model') and os.path.isfile(opt['init_model']):
+            # check first for 'init_model' for loading model from file
+            model_file = opt['init_model']
+
+        if opt.get('model_file') and os.path.isfile(opt['model_file']):
+            # next check for 'model_file', this would override init_model
+            model_file = opt['model_file']
+
+        if model_file is not None:
+            # if we are loading a model, should load its dict too
+            if (os.path.isfile(model_file + '.dict') or
+                    opt['dict_file'] is None):
+                opt['dict_file'] = model_file + '.dict'
+
+        return model_file
+
+    def _set_fixed_candidates(self, shared):
+        if shared:
+            self.fixed_candidates = shared['fixed_candidates']
+            self.fixed_candidates_vec = shared['fixed_candidates_vec']
+        else:
+            if self.opt['fixed_candidates_path']:
+                print("Vectorizing fixed candidates set from {}".format(
+                    self.opt['fixed_candidates_path']))
+                with open(self.opt['fixed_candidates_path'], 'r') as f:
+                    self.fixed_candidates = [line.strip() for line in
+                        f.readlines()]
+                vectorize_func = self._vectorize_text
+                candidate_vecs = [vectorize_func(cand) for cand in
```

Comment:
Actually, unfortunately the truncate flag defaults to -1 and then self.truncate is set to None if truncate < 0, so instead of self.opt.get('truncate'), we need to use self.truncate here. My vote would be to just stick with None throughout; no -1. I assume it's there so we can set the type as `int` for the flag, but we can easily define a one-liner func `int_or_none` that validates the input is int or None and use that instead. -1's as sentinels are scary since they so often work where they shouldn't in python code. What do you think @alexholdenmiller?